File: tests/src/com.google.protobuf/protobuf-javalite/3.25.5/src/test/java/com_google_protobuf/protobuf_javalite/ByteBufferWriterTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import com.google.protobuf.ByteBufferWriterAccess;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Test;

public class ByteBufferWriterTest {
    @Test
    public void directBufferWritePreservesPositionAndCopiesRemainingBytes() throws IOException {
        String prefix = "protobuf ";
        String remaining = "direct byte buffer writer";
        byte[] payload = (prefix + remaining).getBytes(StandardCharsets.UTF_8);
        ByteBuffer buffer = ByteBuffer.allocateDirect(payload.length);
        buffer.put(payload);
        buffer.flip();
        buffer.position(prefix.length());

        ByteArrayOutputStream output = new ByteArrayOutputStream();

        ByteBufferWriterAccess.write(buffer, output);

        assertEquals(prefix.length(), buffer.position());
        assertArrayEquals(remaining.getBytes(StandardCharsets.UTF_8), output.toByteArray());
    }
}

File: tests/src/org.apache.pekko/pekko-protobuf-v3_3/1.0.1/src/test/scala/org_apache_pekko/pekko_protobuf_v3_3/ByteBufferWriterTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_protobuf_v3_3

import java.io.ByteArrayOutputStream
import java.io.OutputStream
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

import org.apache.pekko.protobufv3.internal.ByteString
import org.apache.pekko.protobufv3.internal.UnsafeByteOperations
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ByteBufferWriterTest {
  @Test
  def writesDirectByteBufferBackedByteStringRangeToOutputStream(): Unit = {
    val payload: Array[Byte] = "direct byte buffer backed protobuf data".getBytes(StandardCharsets.UTF_8)
    val directBuffer: ByteBuffer = ByteBuffer.allocateDirect(payload.length)
    directBuffer.put(payload)
    directBuffer.flip()

    val byteString: ByteString = UnsafeByteOperations.unsafeWrap(directBuffer.asReadOnlyBuffer())
    val outputStream: ByteArrayOutputStream = new ByteArrayOutputStream()

    ByteBufferWriterTest.writeRange(byteString, outputStream, 0, byteString.size())

    assertThat(outputStream.toByteArray).containsExactly(payload: _*)
  }
}

object ByteBufferWriterTest {
  private val ByteStringClass: Class[ByteString] = classOf[ByteString]
  private val Lookup: MethodHandles.Lookup = MethodHandles.privateLookupIn(
    ByteStringClass,
    MethodHandles.lookup()
  )

  private val WriteToRange: MethodHandle = Lookup.findVirtual(
    ByteStringClass,
    "writeTo",
    MethodType.methodType(
      java.lang.Void.TYPE,
      classOf[OutputStream],
      java.lang.Integer.TYPE,
      java.lang.Integer.TYPE
    )
  )

  def writeRange(byteString: ByteString, outputStream: OutputStream, offset: Int, length: Int): Unit = {
    WriteToRange.invokeWithArguments(
      byteString,
      outputStream,
      Integer.valueOf(offset),
      Integer.valueOf(length)
    )
    ()
  }
}
